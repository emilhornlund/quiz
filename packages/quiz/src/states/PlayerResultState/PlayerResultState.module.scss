@import '../../styles/colors';
@import '../../styles/helpers';

.score {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  flex-wrap: nowrap;
  background-color: $white-1-50;
  font-weight: bold;

  @mixin score-dimensions($size, $layout-gutter) {
    padding: calc($layout-gutter * 0.5) calc($layout-gutter * 2);
    border-radius: calc($size / 2);
    line-height: $size * 0.7;

    @include fontSize($size * 0.7);
  }

  @include mobile {
    @include score-dimensions(42px, $mobile-narrow-layout-gutter);
  }

  @include tablet {
    @include score-dimensions(52px, $tablet-narrow-layout-gutter);
  }

  @include desktop {
    @include score-dimensions(62px, $desktop-narrow-layout-gutter);
  }
}
