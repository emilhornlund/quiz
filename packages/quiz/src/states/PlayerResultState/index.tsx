export type { PlayerResultStateProps } from './PlayerResultState'
export { default } from './PlayerResultState'
