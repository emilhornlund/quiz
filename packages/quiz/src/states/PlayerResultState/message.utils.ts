export const POSITION_MESSAGES = {
  1: {
    correct: [
      "You're on the podium! First place, and you nailed it!",
      'Champion! Top spot and a perfect answer!',
      "You're number one with the correct answer! Impressive!",
    ],
    incorrect: [
      "You're on the podium! First place, but you missed this one!",
      "Champion, but a wrong answer! You'll bounce back!",
      "Top spot, but this answer wasn't correct!",
    ],
  },
  2: {
    correct: [
      "You're on the podium! So close, second place with a correct answer!",
      'Runner-up with a perfect answer! Amazing job!',
      'Second place and you answered correctly!',
    ],
    incorrect: [
      "You're on the podium! Second place, but a wrong answer!",
      'Runner-up, but you missed this one!',
      'Second place, but the answer wasn’t right.',
    ],
  },
  3: {
    correct: [
      "You're on the podium! Third place and a correct answer!",
      'A bronze finish with the right answer! Well done!',
      'Third place with a correct answer! Keep it up!',
    ],
    incorrect: [
      "You're on the podium! Third place, but you missed it!",
      'A bronze finish, but wrong answer this time!',
      'Third place, but the answer was incorrect.',
    ],
  },
  defaultTop10: {
    correct: [
      'Great effort! Top 10 and you answered correctly!',
      'Top 10 finish with a correct answer! Keep pushing!',
      'You’re in the top 10 and nailed the answer!',
    ],
    incorrect: [
      'Top 10 finish, but you missed the answer!',
      "You're in the top 10, but the answer wasn’t right!",
      "Great effort, but this answer wasn't correct.",
    ],
  },
  defaultTop20: {
    correct: [
      'Good job! Top 20 and you got it right!',
      "Top 20 finish and a correct answer! You're doing great!",
      'Solid performance, and you answered correctly!',
    ],
    incorrect: [
      'Top 20 finish, but the answer wasn’t correct!',
      'Good job, but this one was wrong!',
      'Solid performance, but you missed the answer.',
    ],
  },
  defaultBelow20: {
    correct: [
      'Every game is a learning experience! You got it right this time!',
      'Keep pushing! You got the correct answer!',
      'Great effort! And you answered correctly!',
    ],
    incorrect: [
      'Every game is a learning experience! Keep trying!',
      "Keep playing, and don't worry about this one!",
      "Great effort! You'll get it right next time!",
    ],
  },
}

const getRandomMessage = (messages: string[]) => {
  const randomIndex = Math.floor(Math.random() * messages.length)
  return messages[randomIndex]
}

export const getPositionMessage = (
  position: number,
  answeredCorrectly: boolean,
) => {
  const answerType = answeredCorrectly ? 'correct' : 'incorrect'

  if (position === 1) {
    return getRandomMessage(POSITION_MESSAGES[1][answerType])
  } else if (position === 2) {
    return getRandomMessage(POSITION_MESSAGES[2][answerType])
  } else if (position === 3) {
    return getRandomMessage(POSITION_MESSAGES[3][answerType])
  } else if (position > 3 && position <= 10) {
    return getRandomMessage(POSITION_MESSAGES.defaultTop10[answerType])
  } else if (position > 10 && position <= 20) {
    return getRandomMessage(POSITION_MESSAGES.defaultTop20[answerType])
  } else {
    return getRandomMessage(POSITION_MESSAGES.defaultBelow20[answerType])
  }
}
