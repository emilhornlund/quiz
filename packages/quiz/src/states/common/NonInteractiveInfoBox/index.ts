export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export { default } from './NonInteractiveInfoBox'
