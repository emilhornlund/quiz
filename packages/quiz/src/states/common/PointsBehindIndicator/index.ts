export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export { default } from './PointsBehindIndicator'
