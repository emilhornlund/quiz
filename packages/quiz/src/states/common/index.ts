export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export type { QuestionAnswerPickerProps } from './QuestionAnswerPicker'
export type { QuestionMediaProps } from './QuestionMedia'
export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'

export { default as GamePage } from './GamePage'
export { default as NonInteractiveInfoBox } from './NonInteractiveInfoBox'
export { default as PointsBehindIndicator } from './PointsBehindIndicator'
export { default as QuestionAnswerPicker } from './QuestionAnswerPicker'
export { default as QuestionMedia } from './QuestionMedia'
export { default as QuestionTextPreview } from './QuestionTextPreview'
export { default as QuestionTypePointsBar } from './QuestionTypePointsBar'
