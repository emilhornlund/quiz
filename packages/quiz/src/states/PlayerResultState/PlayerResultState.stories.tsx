import { GameEventType } from '@quiz/common'
import type { Meta, StoryObj } from '@storybook/react'

import tState from './PlayerResultState'

const meta = {
  component: tState,
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof tState>

export default meta
type Story = StoryObj<typeof meta>

export const Correct = {
  args: {
    event: {
      type: GameEventType.GameResultPlayer,
      player: {
        nickname: 'FrostyBear',
        score: {
          correct: true,
          last: 634,
          total: 10458,
          position: 1,
          streak: 3,
        },
      },
      pagination: {
        current: 1,
        total: 20,
      },
    },
  },
} satisfies Story

export const Incorrect = {
  args: {
    event: {
      type: GameEventType.GameResultPlayer,
      player: {
        nickname: 'FrostyBear',
        score: {
          correct: false,
          last: 0,
          total: 10458,
          position: 1,
          streak: 0,
        },
      },
      pagination: {
        current: 1,
        total: 20,
      },
    },
  },
} satisfies Story
