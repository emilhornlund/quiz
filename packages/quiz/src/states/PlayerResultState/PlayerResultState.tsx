import { faCheck, faXmark } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { GameResultPlayerEvent } from '@quiz/common'
import React, { FC, useMemo } from 'react'

import {
  Badge,
  PlayerGameFooter,
  StreakBadge,
  Typography,
} from '../../components'
import { GamePage, PointsBehindIndicator } from '../common'

import { getPositionMessage } from './messages.ts'
import styles from './PlayerResultState.module.scss'

export interface PlayerResultStateProps {
  event: GameResultPlayerEvent
}

const PlayerResultState: FC<PlayerResultStateProps> = ({
  event: {
    player: {
      nickname,
      score: { correct, last: lastScore, total: totalScore, position, streak },
      behind,
    },
    pagination: { current: currentQuestion, total: totalQuestions },
  },
}) => {
  const message = useMemo(() => {
    return getPositionMessage(position, correct)
  }, [position, correct])

  return (
    <GamePage
      footer={
        <PlayerGameFooter
          currentQuestion={currentQuestion}
          totalQuestions={totalQuestions}
          nickname={nickname}
          totalScore={totalScore}
        />
      }>
      <Typography variant="subtitle">
        {correct ? 'Correct' : 'Incorrect'}
      </Typography>
      <Badge size="large" backgroundColor={correct ? 'green' : 'red'}>
        <FontAwesomeIcon icon={correct ? faCheck : faXmark} />
      </Badge>
      <StreakBadge streak={streak}>Streak</StreakBadge>
      <div className={styles.score}>{lastScore}</div>
      <Typography variant="text" size="small">
        {message}
      </Typography>
      {behind && <PointsBehindIndicator {...behind} />}
    </GamePage>
  )
}

export default PlayerResultState
