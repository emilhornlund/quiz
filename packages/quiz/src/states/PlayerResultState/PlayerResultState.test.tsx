import { GameEventType } from '@quiz/common'
import { render } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { describe, it } from 'vitest'

import PlayerResultState from './PlayerResultState'

describe('PlayerResultState', () => {
  it('should render PlayerResultState with default props', async () => {
    render(
      <MemoryRouter>
        <PlayerResultState
          event={{
            type: GameEventType.GameResultPlayer,
            player: {
              nickname: 'FrostyBear',
              score: {
                correct: true,
                last: 634,
                total: 10458,
                position: 1,
                streak: 3,
              },
            },
            pagination: {
              current: 1,
              total: 20,
            },
          }}
        />
      </MemoryRouter>,
    )
  })
})
