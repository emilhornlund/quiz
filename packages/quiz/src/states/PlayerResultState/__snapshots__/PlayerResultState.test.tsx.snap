// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerResultState > renders correct result with behind indicator 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <div
          class="buttonInputContainer buttonInputKindPrimary buttonInputSizeSmall"
        >
          <button
            data-testid="test-fullscreen-button-button"
            id="fullscreen-button"
            name="fullscreen-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-maximize"
              data-icon="maximize"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M168 32L24 32C10.7 32 0 42.7 0 56L0 200c0 9.7 5.8 18.5 14.8 22.2S34.1 223.8 41 217l40-40 79 79-79 79-40-40c-6.9-6.9-17.2-8.9-26.2-5.2S0 302.3 0 312L0 456c0 13.3 10.7 24 24 24l144 0c9.7 0 18.5-5.8 22.2-14.8s1.7-19.3-5.2-26.2l-40-40 79-79 79 79-40 40c-6.9 6.9-8.9 17.2-5.2 26.2S270.3 480 280 480l144 0c13.3 0 24-10.7 24-24l0-144c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2l-40 40-79-79 79-79 40 40c6.9 6.9 17.2 8.9 26.2 5.2S448 209.7 448 200l0-144c0-13.3-10.7-24-24-24L280 32c-9.7 0-18.5 5.8-22.2 14.8S256.2 66.1 263 73l40 40-79 79-79-79 40-40c6.9-6.9 8.9-17.2 5.2-26.2S177.7 32 168 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="typography subtitle full"
          >
            Correct
          </div>
          <div
            class="badge large green"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-check"
              data-icon="check"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M434.8 70.1c14.3 10.4 17.5 30.4 7.1 44.7l-256 352c-5.5 7.6-14 12.3-23.4 13.1s-18.5-2.7-25.1-9.3l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l101.5 101.5 234-321.7c10.4-14.3 30.4-17.5 44.7-7.1z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="streakBadge"
          >
            Streak
            <div
              class="badge small orange"
            >
              3
            </div>
          </div>
          <div
            class="score"
          >
            634
          </div>
          <div
            class="typography text small"
          >
            Message for 1 - correct
          </div>
          <div
            class="pointsBehindIndicator"
          >
            <strong>
              123
            </strong>
             points behind 
            <strong>
              WhiskerFox
            </strong>
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="main"
      >
        <div
          class="questions"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-circle-question"
            data-icon="circle-question"
            data-prefix="fas"
            role="img"
            viewBox="0 0 512 512"
          >
            <path
              d="M256 512a256 256 0 1 0 0-512 256 256 0 1 0 0 512zm0-336c-17.7 0-32 14.3-32 32 0 13.3-10.7 24-24 24s-24-10.7-24-24c0-44.2 35.8-80 80-80s80 35.8 80 80c0 47.2-36 67.2-56 74.5l0 3.8c0 13.3-10.7 24-24 24s-24-10.7-24-24l0-8.1c0-20.5 14.8-35.2 30.1-40.2 6.4-2.1 13.2-5.5 18.2-10.3 4.3-4.2 7.7-10 7.7-19.6 0-17.7-14.3-32-32-32zM224 368a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
              fill="currentColor"
            />
          </svg>
          <span>
            1
             / 
            20
          </span>
        </div>
        <div
          class="nickname"
        >
          <span>
            FrostyBear
          </span>
        </div>
        <div
          class="score"
        >
          <span>
            10458
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PlayerResultState > renders incorrect result without behind indicator 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <div
          class="buttonInputContainer buttonInputKindPrimary buttonInputSizeSmall"
        >
          <button
            data-testid="test-fullscreen-button-button"
            id="fullscreen-button"
            name="fullscreen-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-maximize"
              data-icon="maximize"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M168 32L24 32C10.7 32 0 42.7 0 56L0 200c0 9.7 5.8 18.5 14.8 22.2S34.1 223.8 41 217l40-40 79 79-79 79-40-40c-6.9-6.9-17.2-8.9-26.2-5.2S0 302.3 0 312L0 456c0 13.3 10.7 24 24 24l144 0c9.7 0 18.5-5.8 22.2-14.8s1.7-19.3-5.2-26.2l-40-40 79-79 79 79-40 40c-6.9 6.9-8.9 17.2-5.2 26.2S270.3 480 280 480l144 0c13.3 0 24-10.7 24-24l0-144c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2l-40 40-79-79 79-79 40 40c6.9 6.9 17.2 8.9 26.2 5.2S448 209.7 448 200l0-144c0-13.3-10.7-24-24-24L280 32c-9.7 0-18.5 5.8-22.2 14.8S256.2 66.1 263 73l40 40-79 79-79-79 40-40c6.9-6.9 8.9-17.2 5.2-26.2S177.7 32 168 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="typography subtitle full"
          >
            Incorrect
          </div>
          <div
            class="badge large red"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-xmark"
              data-icon="xmark"
              data-prefix="fas"
              role="img"
              viewBox="0 0 384 512"
            >
              <path
                d="M55.1 73.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L147.2 256 9.9 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192.5 301.3 329.9 438.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.8 256 375.1 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192.5 210.7 55.1 73.4z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="score"
          >
            50
          </div>
          <div
            class="typography text small"
          >
            Message for 7 - incorrect
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="main"
      >
        <div
          class="questions"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-circle-question"
            data-icon="circle-question"
            data-prefix="fas"
            role="img"
            viewBox="0 0 512 512"
          >
            <path
              d="M256 512a256 256 0 1 0 0-512 256 256 0 1 0 0 512zm0-336c-17.7 0-32 14.3-32 32 0 13.3-10.7 24-24 24s-24-10.7-24-24c0-44.2 35.8-80 80-80s80 35.8 80 80c0 47.2-36 67.2-56 74.5l0 3.8c0 13.3-10.7 24-24 24s-24-10.7-24-24l0-8.1c0-20.5 14.8-35.2 30.1-40.2 6.4-2.1 13.2-5.5 18.2-10.3 4.3-4.2 7.7-10 7.7-19.6 0-17.7-14.3-32-32-32zM224 368a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
              fill="currentColor"
            />
          </svg>
          <span>
            2
             / 
            20
          </span>
        </div>
        <div
          class="nickname"
        >
          <span>
            Radar
          </span>
        </div>
        <div
          class="score"
        >
          <span>
            9000
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
